src/plugins/part.c

```diff
@@ -827,38 +826,71 @@ static BDPartSpec** get_disk_parts (const gchar *disk, gboolean parts, gboolean
         }
     }
 
-    num_parts = fdisk_table_get_nents (table);
-    if (num_parts == 0) {
+    /* sort partitions by start */
+    status = fdisk_table_sort_partitions (table, fdisk_partition_cmp_start);
+    if (status != 0) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to sort partitions");
         fdisk_free_iter (itr);
         fdisk_unref_table (table);
         close_context (cxt);
         return NULL;
     }
 
-    ret = g_new0 (BDPartSpec*, num_parts + 1);
-    i = 0;
+    array = g_ptr_array_new ();
 
     while (fdisk_table_next_partition(table, itr, &pa) == 0) {
         spec = get_part_spec_fdisk (cxt, pa, error);
         if (!spec) {
-            for (BDPartSpec **part_list_p = ret; *part_list_p; part_list_p++)
-                bd_part_spec_free (*part_list_p);
-            g_free (ret);
+            g_ptr_array_foreach (array, (GFunc) (void *) bd_part_spec_free, NULL);
+            g_ptr_array_free (array, TRUE);
             fdisk_free_iter (itr);
             fdisk_unref_table (table);
             close_context (cxt);
             return NULL;
         }
 
-        ret[i++] = spec;
+        /* libfdisk doesn't have a special partition for metadata so we need to add
+           a special metadata partition to the "empty" spaces between partitions
+           and free spaces to mimic behaviour of parted
+           metadata partitions should be present in the extended partition in front
+           of every logical partition */
+        if (prev_spec && metadata) {
+            if ((spec->start > prev_spec->start + prev_spec->size) ||
+                (prev_spec->type == BD_PART_TYPE_EXTENDED && spec->start > prev_spec->start) ) {
+                BDPartSpec *ext_meta = g_new0 (BDPartSpec, 1);
+                ext_meta->flags = 0;
```

Comment:
You know `g_new0()` initializes everything to zero, right?

Reply:
Yes, I know. I wanted to set these explicitly here to emphasize that the partition without things like path and name is intention, not a mistake. I can remove it if you think it's not necessary.

Reply:
OK, fair enough.